// @salesforce/salesforcedx-sobjects-faux-generator/out/src/transformer/transformerFactory.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
/*
 * Copyright (c) 2021, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
const core_1 = require("@salesforce/core");
const generator_1 = require("../generator");
const configUtil_1 = require("../generator/configUtil");
const soqlMetadataGenerator_1 = require("../generator/soqlMetadataGenerator");
const retriever_1 = require("../retriever");
const types_1 = require("../types");
const sobjectTransformer_1 = require("./sobjectTransformer");
class SObjectTransformerFactory {
    static create(emitter, cancellationToken, projectPath, category, source) {
        return __awaiter(this, void 0, void 0, function* () {
            const retrievers = [];
            const generators = [];
            const standardGenerator = new generator_1.FauxClassGenerator(types_1.SObjectCategory.STANDARD, 'standardObjects');
            const customGenerator = new generator_1.FauxClassGenerator(types_1.SObjectCategory.CUSTOM, 'customObjects');
            if (source === types_1.SObjectRefreshSource.StartupMin) {
                retrievers.push(new retriever_1.MinObjectRetriever());
                generators.push(standardGenerator);
            }
            else {
                const connection = yield SObjectTransformerFactory.createConnection(projectPath);
                retrievers.push(new retriever_1.OrgObjectRetriever(connection), new retriever_1.OrgObjectDetailRetriever(connection, new GeneralSObjectSelector(category, source)));
                if (category === types_1.SObjectCategory.STANDARD ||
                    category === types_1.SObjectCategory.ALL) {
                    generators.push(standardGenerator);
                }
                if (category === types_1.SObjectCategory.CUSTOM ||
                    category === types_1.SObjectCategory.ALL) {
                    generators.push(customGenerator);
                }
            }
            generators.push(new generator_1.TypingGenerator());
            generators.push(new soqlMetadataGenerator_1.SOQLMetadataGenerator(category));
            return new sobjectTransformer_1.SObjectTransformer(emitter, retrievers, generators, cancellationToken);
        });
    }
    static createConnection(projectPath) {
        return __awaiter(this, void 0, void 0, function* () {
            const connection = yield core_1.Connection.create({
                authInfo: yield core_1.AuthInfo.create({
                    username: yield configUtil_1.ConfigUtil.getUsername(projectPath)
                })
            });
            return connection;
        });
    }
}
exports.SObjectTransformerFactory = SObjectTransformerFactory;
class GeneralSObjectSelector {
    constructor(category, source) {
        this.category = category;
        this.source = source;
    }
    select(sobject) {
        const isCustomObject = sobject.custom === true && this.category === types_1.SObjectCategory.CUSTOM;
        const isStandardObject = sobject.custom === false && this.category === types_1.SObjectCategory.STANDARD;
        if (this.category === types_1.SObjectCategory.ALL &&
            this.source === types_1.SObjectRefreshSource.Manual) {
            return true;
        }
        else if (this.category === types_1.SObjectCategory.ALL &&
            (this.source === types_1.SObjectRefreshSource.StartupMin ||
                this.source === types_1.SObjectRefreshSource.Startup) &&
            this.isRequiredSObject(sobject.name)) {
            return true;
        }
        else if ((isCustomObject || isStandardObject) &&
            this.source === types_1.SObjectRefreshSource.Manual &&
            this.isRequiredSObject(sobject.name)) {
            return true;
        }
        return false;
    }
    isRequiredSObject(sobject) {
        // Ignore all sobjects that end with Share or History or Feed or Event
        return !/Share$|History$|Feed$|.+Event$/.test(sobject);
    }
}
exports.GeneralSObjectSelector = GeneralSObjectSelector;
