// @salesforce/salesforcedx-sobjects-faux-generator/out/src/generator/soqlMetadataGenerator.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
/*
 * Copyright (c) 2021, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
const fs = require("fs");
const path = require("path");
const shelljs_1 = require("shelljs");
const constants_1 = require("../constants");
const messages_1 = require("../messages");
const types_1 = require("../types");
const BASE_FOLDER = [constants_1.TOOLS_DIR, constants_1.SOQLMETADATA_DIR];
class SOQLMetadataGenerator {
    constructor(category) {
        this.category = category;
    }
    generate(output) {
        const outputFolderPath = path.join(output.sfdxPath, ...BASE_FOLDER);
        if (!this.resetOutputFolder(outputFolderPath, this.category)) {
            throw messages_1.nls.localize('no_sobject_output_folder_text', outputFolderPath);
        }
        this.generateTypesNames(outputFolderPath, output.getTypeNames());
        const sobjects = [...output.getStandard(), ...output.getCustom()];
        for (const sobj of sobjects) {
            if (sobj.name) {
                this.generateMetadataForSObject(outputFolderPath, sobj);
            }
        }
    }
    generateTypesNames(folderPath, typeNames) {
        if (!fs.existsSync(folderPath)) {
            fs.mkdirSync(folderPath, { recursive: true });
        }
        const typeNameFile = path.join(folderPath, 'typeNames.json');
        if (fs.existsSync(typeNameFile)) {
            fs.unlinkSync(typeNameFile);
        }
        fs.writeFileSync(typeNameFile, JSON.stringify(typeNames, null, 2), {
            mode: 0o444
        });
    }
    generateMetadataForSObject(folderPath, sobject) {
        if (!fs.existsSync(folderPath)) {
            fs.mkdirSync(folderPath);
        }
        const targetPath = path.join(folderPath, sobject.custom ? constants_1.CUSTOMOBJECTS_DIR : constants_1.STANDARDOBJECTS_DIR, `${sobject.name}.json`);
        fs.writeFileSync(targetPath, JSON.stringify(sobject, null, 2), {
            mode: 0o444
        });
    }
    resetOutputFolder(outputFolder, category) {
        return __awaiter(this, void 0, void 0, function* () {
            const customsFolder = path.join(outputFolder, constants_1.CUSTOMOBJECTS_DIR);
            const standardsFolder = path.join(outputFolder, constants_1.STANDARDOBJECTS_DIR);
            if ([types_1.SObjectCategory.ALL, types_1.SObjectCategory.STANDARD].includes(category) &&
                fs.existsSync(standardsFolder)) {
                shelljs_1.rm('-rf', standardsFolder);
            }
            if ([types_1.SObjectCategory.ALL, types_1.SObjectCategory.CUSTOM].includes(category) &&
                fs.existsSync(customsFolder)) {
                shelljs_1.rm('-rf', customsFolder);
            }
            if (!fs.existsSync(customsFolder)) {
                shelljs_1.mkdir('-p', customsFolder);
            }
            if (!fs.existsSync(standardsFolder)) {
                shelljs_1.mkdir('-p', standardsFolder);
            }
            return true;
        });
    }
}
exports.SOQLMetadataGenerator = SOQLMetadataGenerator;
