// @salesforce/salesforcedx-sobjects-faux-generator/out/src/retriever/orgObjectRetriever.js
"use strict";
/*
 * Copyright (c) 2021, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const describe_1 = require("../describe");
const messages_1 = require("../messages");
class OrgObjectRetriever {
    constructor(connection) {
        this.describer = new describe_1.SObjectDescribe(connection);
    }
    retrieve(output) {
        return __awaiter(this, void 0, void 0, function* () {
            let sobjects = [];
            try {
                sobjects = yield this.describer.describeGlobal();
            }
            catch (e) {
                const err = JSON.parse(e);
                output.setError(messages_1.nls.localize('failure_fetching_sobjects_list_text', err.message), err.stack);
                return;
            }
            output.addTypeNames(sobjects);
        });
    }
}
exports.OrgObjectRetriever = OrgObjectRetriever;
class OrgObjectDetailRetriever {
    constructor(connection, selector) {
        this.describer = new describe_1.SObjectDescribe(connection);
        this.selector = selector;
    }
    retrieve(output) {
        return __awaiter(this, void 0, void 0, function* () {
            let fetchedSObjects = [];
            const retrieveTypes = this.selectedTypes(output.getTypeNames());
            try {
                fetchedSObjects = yield this.describer.fetchObjects(retrieveTypes);
            }
            catch (errorMessage) {
                output.setError(messages_1.nls.localize('failure_in_sobject_describe_text', errorMessage));
                return;
            }
            const standardSObjects = [];
            const customSObjects = [];
            // tslint:disable-next-line:prefer-for-of
            for (let i = 0; i < fetchedSObjects.length; i++) {
                if (fetchedSObjects[i].custom) {
                    customSObjects.push(fetchedSObjects[i]);
                }
                else {
                    standardSObjects.push(fetchedSObjects[i]);
                }
            }
            output.addStandard(standardSObjects);
            output.addCustom(customSObjects);
        });
    }
    selectedTypes(types) {
        return types.reduce((acc, sobject) => {
            if (this.selector.select(sobject)) {
                acc.push(sobject.name);
            }
            return acc;
        }, []);
    }
}
exports.OrgObjectDetailRetriever = OrgObjectDetailRetriever;
